accounting/filters/OsgScheddGpuFilter.py

```diff
@@ -0,0 +1,954 @@
+        if schedd not in self.schedd_collector_host_map_checked:
+            self.logger.debug(f"Schedd {schedd} not found/needs updating in cached collector host map, querying collector")
+            self.schedd_collector_host_map[schedd] = set()
+            self.schedd_collector_host_map_checked.add(schedd)
+            new_hosts = False
+
+            collectors_queried = set()
+            for collector_host in self.collector_hosts:
+                if collector_host in {"flock.opensciencegrid.org"}:
+                    continue
+                collector = htcondor.Collector(collector_host)
+                try:
+                    ads = collector.query(
+                        htcondor.AdTypes.Schedd,
+                        constraint=f'''Machine == "{schedd.split('@')[-1]}"''',
+                        projection=["CollectorHost"],
+                    )
+                except htcondor.HTCondorIOError:
+                    continue
+                collectors_queried.add(collector_host)
+                ads = list(ads)
+                if len(ads) == 0:
+                    continue
+                if len(ads) > 1:
+                    self.logger.warning(f'Got multiple Schedd ClassAds for Machine == "{schedd}"')
+
+                # Cache the CollectorHost in the map
+                if "CollectorHost" in ads[0]:
+                    schedd_collector_hosts = set()
+                    for schedd_collector_host in ads[0]["CollectorHost"].split(","):
+                        schedd_collector_host = schedd_collector_host.strip().split(":")[0]
+                        if schedd_collector_host:
+                            schedd_collector_hosts.add(schedd_collector_host)
+                            new_hosts = True
+                    if schedd_collector_hosts:
+                        self.schedd_collector_host_map[schedd] = schedd_collector_hosts
+                        break
+            else:
+                self.logger.warning(f"Did not find Machine == {schedd} in collectors {', '.join(collectors_queried)}")
+
+            # Update the pickle
+            if new_hosts and len(schedd_collector_hosts) > 0:
+                self.logger.debug(f"Updating collector host pickle for {schedd} with {schedd_collector_hosts}")
+                old_schedd_collector_host_map = {}
+                if self.schedd_collector_host_map_pickle.exists():
+                    try:
+                        old_schedd_collector_host_map = pickle.load(open(self.schedd_collector_host_map_pickle, "rb"))
+                        old_schedd_collector_host_map[schedd] = schedd_collector_hosts
+                    except IOError:
+                        pass
+                old_schedd_collector_host_map[schedd] = schedd_collector_hosts
+                with open(self.schedd_collector_host_map_pickle, "wb") as f:
+                    pickle.dump(old_schedd_collector_host_map, f)
+
+        return self.schedd_collector_host_map[schedd]
```

Comment:
I would like to replace this block with what happens in all of the other OSPool reports (except for the "main" CPU one). The idea is that the main report runs first and does all the updating of the CollectorHost dict/pickle and so the rest of the reports shouldn't need to do that. Most likely, all the schedds' CollectorHosts should already be cached. See e.g. https://github.com/CHTC/JobAccounting/blob/412e53e09495c7008bb16771d9525bb2e59575df/accounting/filters/OsgScheddCpuRemovedFilter.py#L99-L137 (which is where this diff comes from)
```suggestion
    def schedd_collector_host(self, schedd):
        # Query Schedd ad in Collector for its CollectorHost,
        # unless result previously cached
        if schedd not in self.schedd_collector_host_map:
            self.schedd_collector_host_map[schedd] = set()

            collectors_queried = set()
            for collector_host in self.collector_hosts:
                if collector_host in {"flock.opensciencegrid.org"}:
                    continue
                collector = htcondor.Collector(collector_host)
                try:
                    ads = collector.query(
                        htcondor.AdTypes.Schedd,
                        constraint=f'''Machine == "{schedd.split('@')[-1]}"''',
                        projection=["CollectorHost"],
                    )
                except htcondor.HTCondorIOError:
                    continue
                collectors_queried.add(collector_host)
                if len(ads) == 0:
                    continue
                if len(ads) > 1:
                    self.logger.warning(f'Got multiple Schedd ClassAds for Machine == "{schedd}"')

                # Cache the CollectorHost in the map
                if "CollectorHost" in ads[0]:
                    schedd_collector_hosts = set()
                    for schedd_collector_host in ads[0]["CollectorHost"].split(","):
                        schedd_collector_host = schedd_collector_host.strip().split(":")[0]
                        if schedd_collector_host:
                            schedd_collector_hosts.add(schedd_collector_host)
                    if schedd_collector_hosts:
                        self.schedd_collector_host_map[schedd] = schedd_collector_hosts
                        break
            else:
                self.logger.warning(f"Did not find Machine == {schedd} in collectors {', '.join(collectors_queried)}")

        return self.schedd_collector_host_map[schedd]
```